### ISSUE TITLE
Decide on namespace and naming

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph-cpp#3]
We need a decision on the namespace and symbol naming. The three issues to make a decision on are:

 - Currently the namespace is `ig` in order to avoid conflict with the `igraph` namespace of https://github.com/ntamas/igraphpp/

 - Additionally, all symbol names start with `ig`. This is indeed redundant, but I find it more tedious to type `ig::Vec` than `igVec`. One of the major goals is to make igraph easier and more pleasant to use. This involves shortening names.

 - It was suggested to use `vector` or `Vector` instead of `Vec`. These are more conflict-prone and probably rule out omitting the namespace (`using namespace ig`) in the majority of realistic use cases.

I don't have strong opinions on any of these issues. I would like to go with whatever most people are happy with.

Let me know your opinions @ntamas @Antonov548 @vtraag (you might use this library in `leidenalg`) @LTLA

**Comment by LTLA** [comment]
My 2c: the length of the name of the type has rarely been an issue for me. I'm happy to use longer names - within reason, I don't code in Java - provided that I can easily understand the origin and nature of the type when I read the code. 

So, for example, something like `igraph::IntegerVector` is unambiguous and easy to understand - it's an integer vector from `igraph` (or some equivalently descriptive namespace, if we want to keep it separate from **igraphpp**), no need to consult documentation at all. Currently, the name is written as `ig::igIntVec`; a new developer could probably guess what this is, but I would rather spend the extra few characters to just spell it out in plain English.

I would hypothesize that the ease of use of a library is more determined by how easy it is to understand, not necessarily how easy it is to type (again, within reason). A slightly longer but self-documenting name would be helpful for the former.

Oh, and FWIW, I always include the namespace in my type declarations. I stopped `using namespace <xxx>` about a decade ago when I shot off my own foot from writing a function with the same name as something in `std::`.

**Comment by ntamas** [comment]
> One of the major goals is to make igraph easier and more pleasant to use. This involves shortening names.

I don't agree with this either. Most IDEs have autocompletion so shortening `Vector` to `Vec` only hurts readability but does not really save a significant amount of keystrokes. I'd rather spell out everything in full, within reason.

Re namespaces: `using namespace xxx;` feels akin to `from xxx import *` in Python, which is in general discouraged as IDEs can not track easily where a particular symbol comes from without looking at the context. I'm not using this construct in my C++ code either.

So, my suggestion:

* Using `ig` as the namespace is okay; people have adopted this in Python as well so it kinda makes sense. But `igraph` is fine for me as well, no one would want to use my old C++ wrapper together with this one anyway.
* I would not prefix the symbol names with `ig` -- it defeats the whole purpose of a namespace. Why do we have a namespace if we prefix everything with the same string in the namespace itself?
* `vector` would probably be problematic in C++ due to the possible clash with the STL. `Vector` seems okay for me. I can tolerate shortening `Vector` to `Vec` and `Matrix` to `Mat` if it has to, but let's not push this to the extreme. I'd rather have longer but human-readable names. Also, I'd aim for consistency with the names in the C core. I guess we won't document everything thoroughly in the C++ interface, in which case it makes it easier to connect things back to the documentation of the C core if the functions and types are named the same way, or if you can trace back the C name of a function or type from the C++ name with a short and unambiguous rule set (like: translate from CamelCase to underscore_case, snap `igraph_` in front and there you go).

**Comment by szhorvat** [comment]
I removed `ig`, kept vector as `Vec` and matrix as `Mat`, but I'm not averse to spelling these out in full if people insist. I don't want to spell out `Int` as `Integer` though. (I'd rather add `igraph_int_t` as an alias to `igraph_integer_t` to igraph if we want consistency, after all we have `igraph_uint_t` as well.)
